packages/contracts/src/dollar/libraries/LibDollarAmoMinter.sol

```diff
@@ -0,0 +1,473 @@
+import "../interfaces/IUbiquityDollarToken.sol";
+import "../interfaces/IUbiquityCreditToken.sol";
+import "../interfaces/IAMO.sol";
+import "./Constants.sol";
+import "../utils/TransferHelper.sol";
+
+library LibDollarAmoMinter {
+    bytes32 constant AMOMINTER_CONTROL_STORAGE_SLOT =
+        bytes32(uint256(keccak256("ubiquity.contracts.amominter.storage")) - 1);
+
+    /* ========== EVENTS ========== */
+
+    event AMOAdded(address amo_address);
+    event AMORemoved(address amo_address);
+    event Recovered(address token, uint256 amount);
+
+    /* ========== STATE VARIABLES ========== */
+    struct DollarAmoMinterData {
+        // Core
+        IUbiquityDollarToken UAD;
+        IUbiquityCreditToken UCR;
+        ERC20 collateral_token;
+        IUbiquityPool pool;
+        address timelock_address;
+        address custodian_address;
+        // Collateral related
+        address collateral_address;
+        uint256 col_idx;
+        // AMO addresses
+        address[] amos_array;
+        mapping(address => bool) amos; // Mapping is also used for faster verification
+        // Max amount of collateral the contract can borrow from the UbiquityPool
+        int256 collat_borrow_cap;
+        // Max amount of uad and ucr this contract can mint
+        int256 uad_mint_cap;
+        int256 ucr_mint_cap;
+        // Minimum collateral ratio needed for new uad minting
+        uint256 min_cr;
+        // uad mint balances
+        mapping(address => int256) uad_mint_balances; // Amount of FRAX the contract minted, by AMO
+        int256 uad_mint_sum; // Across all AMOs
+        // ucr mint balances
+        mapping(address => int256) ucr_mint_balances; // Amount of FXS the contract minted, by AMO
+        int256 ucr_mint_sum; // Across all AMOs
+        // Collateral borrowed balances
+        mapping(address => int256) collat_borrowed_balances; // Amount of collateral the contract borrowed, by AMO
+        int256 collat_borrowed_sum; // Across all AMOs
+        // uad balance related
+        uint256 uadDollarBalanceStored;
+        // Collateral balance related
+        uint256 missing_decimals;
+        uint256 collatDollarBalanceStored;
+        // AMO balance corrections
+        mapping(address => int256[2]) correction_offsets_amos;
+    }
+
+    /**
+     * @notice Returns struct used as a storage for this library
+     * @return l Struct used as a storage
+     */
+    function dollarAmoMinterStorage()
+        internal
+        pure
+        returns (DollarAmoMinterData storage l)
+    {
+        bytes32 slot = AMOMINTER_CONTROL_STORAGE_SLOT;
+        assembly {
+            l.slot := slot
+        }
+    }
+
+    function init(
+        address _custodian_address,
+        address _timelock_address,
+        address _collateral_address,
+        address _collateral_token,
+        address _pool_address,
+        address _ucr,
+        address _uad
+    ) internal {
+        dollarAmoMinterStorage().uadDollarBalanceStored = 0;
```

Comment:
Please refactor this, you can instance the storage in a var and the re-use the pattern

Reply:
done